### PR TITLE
[VS Incentives] feat: non-perpetual Group & group Gauge pruning

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #6434

## What is the purpose of the change

This PR focuses specifically on non-perpetual Gauge and associated Group pruning.

For context, the end goal is to:
- Write groups to a separate store index
- Write group gauges into a regular gauge index but do not create upcoming/active/finished ref keys
- Pool incentives distribute to group gauges retrieved from the regular index
- In distribution logic, get all groups present, iterate over them and retrieve the associated group gauges
- At the end of the distribution logic, prune non-perpetual expired gauges and groups. If any dust remaining, send to community pool
- Restrict external incentive providers to only be able to create non-perpetual gauges so that it is impossible to create permanent perpetual junk even with the fee
  * This is consistent with the fact that most external are non-perpetual anyways
please correct me if not
- Give subdao powers to remove perpetual gauges that have no use via a special message (alternatively, gov prop)

The pruning is achieved by introducing a new `handleGroupPostDistribute` method that is being called from `AllocateAcrossGauges` at the end of incentive distribution.

## Testing and Verifying

- Adds unit tests for `handleGroupPostDistribute`
- Adds unit tests for `TestIsLastNonPerpetualDistribution`
- Updates unit tests for `AllocateAcrossGauges`

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/incentives/types/keys.go [hunk 1]
```diff
@@ -68,7 +71,7 @@ func NoLockInternalGaugeDenom(poolId uint64) string {
 	return fmt.Sprintf("%s%d", NoLockInternalPrefix, poolId)
 }
 
-// KeyGroupByGaugeID returns key for a given groupGaugeId.
+// KeyGroupByGaugeID returns group key for a given groupGaugeId.
 func KeyGroupByGaugeID(groupGaugeId uint64) []byte {
-	return []byte(fmt.Sprintf("%s%s%d%s", GroupPrefix, "|", groupGaugeId, "|"))
+	return []byte(fmt.Sprintf("%s%d%s", KeyPrefixGroup, groupGaugeId, KeyIndexSeparator))
```
Comment (p0mvn) [inline x/incentives/types/keys.go:76]: Note: I removed one separator as I see no need

File: x/incentives/keeper/distribute.go [hunk 2]
```diff
@@ -115,12 +116,10 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context, activeGroups []types.Group
...
 		}
 
-		// Update total coins distributed and filled epoch of the group gauge.
-		if err = k.updateGaugePostDistribute(ctx, *groupGauge, coinsToDistribute); err != nil {
+		// Either updates the group or deletes it from state if it is finished.
+		if err = k.handleGroupPostDistribute(ctx, *groupGauge, coinsToDistribute); err != nil {
 			return err
 		}
-
-		// TODO: move group from active to finished if applicable.
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:123]: Note: this is obsolete now

File: x/incentives/keeper/distribute_test.go [hunk 3]
```diff
@@ -2051,14 +2061,23 @@ func (s *KeeperTestSuite) TestAllocateAcrossGauges() {
...
-				s.Require().Equal(groupConfig.expectedTotalDistribution, actualDistributed)
+					// Check that the group was deleted
+					_, err = s.App.IncentivesKeeper.GetGroupByGaugeID(s.Ctx, groupConfig.group.GroupGaugeId)
+					s.Require().Error(err)
+					s.Require().ErrorIs(err, types.GroupNotFoundError{GroupGaugeId: groupConfig.group.GroupGaugeId})
+				} else {
+					// check that the group gauge distributed epoch updated
+					s.Require().Equal(groupConfig.groupGauge.FilledEpochs+1, groupGauge.FilledEpochs)
 
-				// TODO: check that group gauge was moved to finished if applicable
```
Comment (p0mvn) [inline x/incentives/keeper/distribute_test.go:2061]: Note: this is obsolete now

File: x/incentives/keeper/gauge.go [hunk 4]
```diff
@@ -227,6 +230,9 @@ func (k Keeper) CreateGroup(ctx sdk.Context, coins sdk.Coins, numEpochPaidOver u
 		return 0, fmt.Errorf("Invalid splitting policy, needs to be Evenly got %s", splittingPolicy)
 	}
 
+	// TODO: remove gauge creation logic from here.
+	// Instead, call `CreateGauge` directly
+	// Update `CreateGauge` to be able to handle the group type.
```
Comment (p0mvn) [inline x/incentives/keeper/gauge.go:235]: Note: tracked https://github.com/osmosis-labs/osmosis/issues/6513

Comment (AlpinYukseloglu) [reply x/incentives/keeper/gauge.go:235]: Had a comment on this but added directly to the issue so it doesn't get lost when we merge this PR

File: x/incentives/keeper/gauge.go [hunk 5]
```diff
@@ -280,7 +286,7 @@ func (k Keeper) GetGaugeByID(ctx sdk.Context, gaugeID uint64) (*types.Gauge, err
 	store := ctx.KVStore(k.storeKey)
 	gaugeKey := gaugeStoreKey(gaugeID)
 	if !store.Has(gaugeKey) {
-		return nil, fmt.Errorf("gauge with ID %d does not exist", gaugeID)
+		return nil, types.GaugeNotFoundError{GaugeID: gaugeID}
```
Comment (AlpinYukseloglu) [inline x/incentives/keeper/gauge.go:289]: Thanks for doing these driveby cleanups!

File: x/incentives/keeper/distribute.go [hunk 6]
```diff
@@ -647,6 +646,39 @@ func (k Keeper) updateGaugePostDistribute(ctx sdk.Context, gauge types.Gauge, ne
...
+// the distributed coins field to add the coins distributed.
+// If group gauge is non-perpetual and at the last distribution epoch, it will delete the group and group gauge from state.
+// Before deleting the non-perpetual gauge, any difference between distributed coins and gauge's coins is sent to
+// community pool.
+// CONTRACT: non-perpetual gauge at the last distribution epoch must have distributed all of its coins already
+func (k Keeper) handleGroupPostDistribute(ctx sdk.Context, groupGauge types.Gauge, coinsDistributed sdk.Coins) error {
+	// Prune expired non-perpetual gauges.
+	if groupGauge.IsLastNonPerpetualDistribution() {
+		// Send truncation dust to community pool.
+		truncationDust := groupGauge.Coins.Sub(groupGauge.DistributedCoins.Add(coinsDistributed...))
```
Comment (AlpinYukseloglu) [inline x/incentives/keeper/distribute.go:660]: Even though it should technically not be possible, I think it would be worthwhile to do a `IsAllGTE` check here to ensure that this dust is never negative so that e.g. weird rounding behavior cannot be escalated to halting the chain

File: x/incentives/keeper/distribute_test.go [hunk 7]
```diff
@@ -2080,3 +2099,121 @@ func (s *KeeperTestSuite) setupVolumes(poolIds []uint64, updatedPoolVolumes []os
...
+			// Note that total coins in gauge = 2x defaultCoins
+			// Already distributed coins = none
+			// Coins to distribute = defaultCoins
+			// Therefore, the community pool should receive 2x defaultCoins - defaultCoins = defaultCoins
+			groupGauge:       withCoinsToDistribute(withNonPerpetualEpochs(withIsPerpetual(deepCopyGauge(defaultPerpetualGauge), false), 0, 1), defaultCoins.Add(defaultCoins...), sdk.NewCoins()),
+			coinsDistributed: defaultCoins,
+			amountToPrefund:  defaultCoins,
+		},
+
+		// // edge cases
```
Comment (AlpinYukseloglu) [inline x/incentives/keeper/distribute_test.go:2134]: Assuming this was intentionally double-commented to make it clear it's a separator? Might be good for us to align on a separator pattern that we all like for consistency (I'm personally indifferent as long as it's roughly consistent across our codebase)

File: x/incentives/keeper/distribute_test.go [hunk 8]
```diff
@@ -2080,3 +2099,121 @@ func (s *KeeperTestSuite) setupVolumes(poolIds []uint64, updatedPoolVolumes []os
...
+
+			if tc.expectError {
+				s.Require().Error(err, "test: %s", name)
+				return
+			}
+			s.Require().NoError(err, "test: %s", name)
+
+			finalIncentivesModuleBalance := s.App.BankKeeper.GetAllBalances(s.Ctx, s.App.AccountKeeper.GetModuleAddress(types.ModuleName))
+
+			if !tc.groupGauge.IsPerpetual && tc.groupGauge.IsLastNonPerpetualDistribution() {
```
Comment (AlpinYukseloglu) [inline x/incentives/keeper/distribute_test.go:2183]: Might be out of scope for this PR, but it would be good to eventually add a test covering a non-perpetual gauge that starts with having e.g. 10 epochs left and then runs 10 epochs of distributions after which we assert that it's empty and cleared from state

**Review by AlpinYukseloglu — APPROVED** [review]
LGTM, nice work

File: x/incentives/keeper/gauge.go (hunk 4 above)
Comment (p0mvn) [reply x/incentives/keeper/gauge.go:235]: Replied in the issue

File: x/incentives/keeper/distribute.go (hunk 6 above)
Comment (p0mvn) [reply x/incentives/keeper/distribute.go]: Addressed: [702bb34](https://github.com/osmosis-labs/osmosis/pull/6509/commits/702bb3497dfa53ec58d9a18a6c612b85d08224c3)

File: x/incentives/keeper/distribute_test.go (hunk 7 above)
Comment (p0mvn) [reply x/incentives/keeper/distribute_test.go]: This was a typo. Addressed. Happy to make an issue to discuss comment consistency in tests

File: x/incentives/keeper/distribute_test.go (hunk 8 above)
Comment (p0mvn) [reply x/incentives/keeper/distribute_test.go:2233]: Added here, good idea: [850b041](https://github.com/osmosis-labs/osmosis/pull/6509/commits/850b0415ff4ed996b1491fc58debe4245c12c766)

File: x/incentives/keeper/distribute.go (hunk 6 above)
Comment (p0mvn) [reply x/incentives/keeper/distribute.go]: I expanded the check further here: [850b041](https://github.com/osmosis-labs/osmosis/pull/6509/commits/850b0415ff4ed996b1491fc58debe4245c12c766)

The first update was correct as well. However, the last one makes it stricter

File: x/incentives/keeper/gauge.go [hunk 9]
```diff
@@ -50,6 +50,9 @@ func (k Keeper) setGauge(ctx sdk.Context, gauge *types.Gauge) error {
 	if err != nil {
 		return err
 	}
+
+	gauge.IsLastNonPerpetualDistribution()
+
```
Comment (p0mvn) [inline x/incentives/keeper/gauge.go:55]: Note: this was accidentally added here. Removed in a future PR
